### ISSUE TITLE
[UX] Simplify/automate the process of creating a website contact form.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1168]
I usually use webform for all my form needs, but I gave core contact.module a go last night after a really long time. Now, I consider myself a relatively drupal/backdrop-savvy user, so that simple task should (in theory) have taken me a minute or two, but it seemed like an endless trip of pogo-sticking around the UI to achieve this:

1. Go to the modules page and enable the contact.module. (but I already knew of its existence because old-time drupal user as already said).
2. All I needed was a simple site-wide contact form. So naturally, after enabling the module I went searching for a setting under Config -> System -> Site info ...nope! nothing there!? :-1:  ...aah! There's a Structure -> Contact form (used admin_bar search field to find it - good UX btw :+1: ). There was a "Website feedback" category there (but what the hell is a category in this context here? UX WTF?) and it already was set to the site's email address as "Recipients", but what the hell is the "Selected: Yes" thing? ...oh well... moving on...
3. Select the "Edit" operation and I got in the "Edit contact category" page. What is this page used for? Where do I actually configure the form? Where do I set its path and menu item? If you are experienced with forms and some of the lingo, maybe you finally understand what this config page is about. I personally would not call these "contact categories" - more likely "contact purpose" or "contact department" or something like that. I don't claim to have found a better alternative to "contact categories" - merely pointing the fact that the terminology does not convey the use and we need to come up with something better. Anyways, point is a) all I need is a very simple contact form for the website - no contact categories, just a quick way for people to send an email to the owner of the website b) I clearly am not in the part of the admin UI that does what I want. Moving on...
4. Looking for tutorials/documentation, I find that once contact.module is enabled, it creates the form under `http://my.site.net/contact`. Cool! I'm getting somewhere, but why not a menu item? Go to add a menu item in the primary navigation and there it is. Good! saved my a couple of steps! :+1: ...but why disabled? Anyways, enable it, move it after the "Home" link (expected it there by default), moving on...
5. Hit that page as anonymous... access denied! WTF!? OK, drupal old-timer goes to permission page and sets permissions for anonymous. All good.

So, list of things that we should discuss about changing:

* Have contact.module enabled by default? Only on the standard profile of course (#1571).
* Have the "Contact" menu item enabled by default when contact.module is enabled (#1572) ...and make sure that it comes after the "Home" link.
* Configuring the site-wide contact form. (https://github.com/backdrop/backdrop-issues/issues/2990) Lets do what @jenlampton has suggested previously for other UI parts and put things where people expect to find them. What I would like to see is either a "use this email for the site-wide contact form" checkbox right below the email in the "Site details" fieldset on the Site Information page, or even better a dedicated "Contact form" fieldset. The dedicated fieldset could have a second email field (perhaps with a "use same as site email above" checkbox for convenience) + a contact form path/URL field + another couple of checkboxes: one to enable a/the "Contact" menu item (if not a checkbox, then at least a link to the primary menu link edit form), another checkbox to expose the contact form as a block and another set of checkboxes to select which roles are allowed to use the form (again, if that's too much, the at least a link to the permissions page ...if possible filtered down to the permission in question. See #980).

**Comment by docwilmot** [comment]
Just had a thought, only slightly related: how about a new feature (two actually) to improve discovery of new menu items?:
- A persistent backdrop_set_message that says "New menu link added! "Contact Forms is now at Structure > Contact Forms."
- (Second feature is the ability to have persistent messages with a "dismiss" link, or even an [X] to make this work.)

**Comment by sutibun** [comment]
I have a few thoughts based on my initial experience.

The long winded story
-----

After finally setting up my first Drupal site as best I could, I thought of adding a contact page with a contact form because I didn't want to provide my email address. Luckily, there was a contact module but it was confusing. No need for me to go into it. Something similar to what klonos mentioned. Basically, the module wasn't self explanatory and in the end it also wasn't exactly what I wanted.

I tried using webform but because of my shared hosting (or something else?) I receive an out of memory error whenever I enabled it. So, currently, I just use Google forms.

So I have real need for this.

The ideal contact form
----

Here is how I envision the simple process to go:

1. Enables a contact module.
2. Module installs and a persistent message appears on the top of the page (swiping docwilmot's idea)
    
    > A contact page has been created at http://example.com/contact Click \_here\_ to add it to the primary menu. You may also edit the page \_here\_.
    > 
    > Alternatively, a standalone contact form block has also been created. Click \_here\_ to decide where to place it in the Layout.
3. If I click to add it to the first menu, I am done. If I choose to go with the standalone contact form, I can choose to place it anywhere I want using the Layout system.
4. Either way, the contact module also enables the Honeypot module by default, in order to protect the site from the sure to come spam. (Now, that's helping the user)

The implementation details
-----

I don't know if this can be done but I think giving both options 

1. creating the contact page with the form in it vs 
2. a standalone contact form you can place on the sidebar or anywhere)

is good enough to cover the majority of cases. I myself have uses for both.

Here is how I imagine this being done on a conceptual level is:

- Modules provides a standalone contact form in the form of a block (instead of calling everything a block, we can call it a dynamic object? or something nicer) so the user can use the Layout system to place it anywhere on the site.
- To create the Contact page, the module simply creates a Basic Page, gives it the default URL /contact, dynamically creates a custom layout where the standalone form is placed under the Contact Page content.

Let's review 
----
The benefit I see of this approach is reusing what Backdrop already has, subtly teaching users how the Layout system works (assuming I understand it as well I think I do. Haven't used it heavily) and more importantly for me, it allows you to easily add your own content to the contact page and customize the URL. (me being nitpicky, I'll probably override the default /contact to /about/contact That's just preference though. My preference shouldn't be the default.)

Most normal users won't know that by enabling forms, they're exposing their sites to spammers so I think Backdrop should **go the extra mile** and enable the Honeypot module by default (users can deactivate them in the config if they so choose). It will make the UX that much better. It'll show the world some thought went behind this instead letting users fend for themselves.

This process I believe is a smart default move.

Personally, I think the contact form should be off by default. Or, maybe a better move would be to ask the user upon installation if they want a contact page/form.

**Comment by klonos** [comment]
> - (Second feature is the ability to have persistent messages with a "dismiss" link, or even an [X] to make this work.)

@docwilmot Yep, same as your suggestion in https://github.com/backdrop/backdrop-issues/issues/1062#issuecomment-123211483:

> You could also port https://www.drupal.org/project/disable_messages. :smile:

I'd like to see that happen.

**Comment by docwilmot** [comment]
> You could also port https://www.drupal.org/project/disable_messages. 

We considered this already, but thought it would need API changes to do it properly for core. See https://github.com/backdrop/backdrop-issues/issues/1062

**Comment by klonos** [comment]
> ...see #1062 

Yes I know. I was following that issue (in fact I follow all backdrop issues).

**Comment by docwilmot** [comment]
I think the problems stated here are all very real. The contact form is not very intuitive. I think though keeping it disabled by default is reasonable, but not so sure. When enabled though, there should be a single point of entry to configure it, as follows.
My suggestion for the main contact form admin page (admin/structure/contact):
- Add information to show where the contact form will be (default `/contact`) and a textfield to change it. Also option to hide this page if using a contact block.
- Add a list of permissions, with checkboxes, to view the contact form
- Add some explanation as to what "categories" are, but keep the categories drag list
- Add an enable checkbox/button at the top/bottom which once toggled turns the contact form on, and adds it to menus, and sorts out permissions, all at once.
- Add contextual links to the contact form itself, to edit categories, or go to the admin page.

We should definitely Include a contact block as well in core, great idea.

How does this soound?

**Comment by sutibun** [comment]
Disabled or Enabled? I can go either way. The only thing is we should have some form of spam protection in place first (#1169 whether AJAX comments, Honeypot, or something else) to protect new users **if enabled by default.**

Those admin changes makes sense to me. I say go for it.

Can't we also change the module's approach as I suggested.(a Basic Page + contact form block + Layout) Why can't we use a Basic page with default url /contact and using the Layout system to append the contact form. Seems much simpler and easier to develop (of course, I'm not going to build it so I can say that easily :p) since you're reusing what Backdrop comes with by default.

The module's current implementation doesn't allow me to easily add content. You either stick with the below look or look for something else.

![contact](https://cloud.githubusercontent.com/assets/14132746/10110601/ddd1ec44-6393-11e5-9649-d11826eb51bb.png)

To me, not helpful... like having to stick with a CMS' default theme.

- If you're a small-medium sized creative agency, there's a need to add [other ways to contact you](http://www.ecommercepartners.net/contact-us) Maybe even a picture of a map.
    
    ![ecp](https://cloud.githubusercontent.com/assets/14132746/10111292/a7244d9a-6398-11e5-8922-8cc1cc894163.png)
- If you're a government agency, you'll want to add [department phone numbers](http://apps.environment-agency.gov.uk/contact/default.aspx)
- In my case, I'd like to add some text before the form. Possibly even some guidelines before contacting the site.

Conceptually, a contact page is just another page and it'll be more user friendly (I think) to see it under "Content > Find content" as a Basic Page type.

If we provide a Basic Page as the basis of the contact page, editing and adding text is much easier. As well as changing the URL.

Another idea, though possibly for the future: besides receiving an email copy, the module can store received messages and display them in a nice table in the admin section.

Also, related to this issue is #569

**Comment by docwilmot** [comment]
Here's the general idea for the admin. Could probably have a link to got to the form too. And move 'Add category to below the category table, andopen category edit in a dialog. That would remove all of the multistep jumping @klonos talked about. 

![contact form localhost](https://cloud.githubusercontent.com/assets/1042283/10113563/43b3da06-63ab-11e5-9c7f-f60de59a3290.png)

**Comment by sutibun** [comment]
Between the label "URL path for contact form" and its field, you can probably provide a link to the actual contact page. I'm not sure if that's what you meant by

> Could probably have a link to got to the form too

if so, never mind then.

**Comment by jenlampton** [comment]
I like where this is headed. 

Contact module as it is now does two things: 1) provide a whole whirlwind of site-wide contact forms. 2) provide a way for users to contact each other through their own personal contact forms.  Neither of these really hits our primary use-case of wanting a single stand-alone contact form (possibly with content at the top) and/or a block that can be placed anywhere.

We can't use a basic page for the content at the top, because not all sites will have that content type.  What happens if they don't?

What about something like this:
- Remove contact module as it is now from core. It's too much.
- Instead, provide a `contact form ` module (naming things are hard)
- Provide a settings form for the contact form that allows you to set:
  * email address where the form should send mail
  * do you want a contact page (yes/no)

If someone clicks YES for wanting a page, we can have the module set up a layout as follows:
- copy the default layout
- create a block, add it at the top of the main content area (with placeholder content?)
- place the contact block below
- sets a message on the admin form saying "You now have a contact page here!" or "Add it to your menu!"

I'm adding the milestone 2.x because replacing the current contact module with something more appropriate will certainly involve API changes.

**Comment by sutibun** [comment]
> We can't use a basic page for the content at the top, because not all sites will have that content type. What happens if they don't?

I thought Backdrop came with "Basic Page" as a default content type, along with "Article". In case it was deleted by the user, not that I would know, but couldn't a module create the content type again? btw, wouldn't those that delete default content types be not normal users?

I just thought the Basic Page route was easier for users.

- **Want to add/edit content on your contact page?** Just go the Basic Page called "Contact" in "admin>content>find content". Since it's no different from editing any other content, users will be familiar with this route., 
- **Want to change the actual contact form?** Go to the modules config page ("admin>structure>contact form"? whatever it is) 

My only problem with the block route is that editing blocks in "admin>structure>custom blocks>list blocks" is not as intuitive (only suspecting) as going to edit a Basic Page. (which they'll have to learn if they want to create content on their website)

basically,

- "admin>structure>custom blocks>list blocks" => huh? what's a block... a brick?
- "admin>content>find content" => makes sense

But in the end, the block route is way better than the current way so I am not totally against it. I guess maybe it's a minor difference: Basic Page vs Block

**Comment by jenlampton** [comment]
You can also edit a block from the layout ;)

**Comment by docwilmot** [comment]
> our primary use-case of wanting a single stand-alone contact form

I thought that was exactly what the current contact form does, no? It has categories to set the email recipient (I'm assuming thats what you mean by "a whole whirlwind of site-wide contact forms") but there is only **one** form, at `/contact` as far as I can tell.

I don't think the contact form itself is terribly complex or too much, in fact I think it is too little. :-). I do agree the personal contact forms are much.

As far as 
> copy the default layout
create a block, add it at the top of the main content area (with placeholder content?)
place the contact block below
sets a message on the admin form saying "You now have a contact page here!" or "Add it to your menu!"

This is great. I like the idea of giving `/contact` its own layout. Can't we in the interim do this with the current contact form and an upgrade script?

**Comment by sutibun** [comment]
If it means the "ideal" contact form won't arrive until much later 2.x (especially since I won't help build it) I'd say lets go with the fix ups from @docwilmot It's the most practical in the meantime and a huge improvement to the current setup.

**Comment by klonos** [comment]
Yeah, lets offer the additional ease of building a contact form in 1.x now and lets discuss removing features in 2.x later.

**Comment by klonos** [comment]
Back to this...

> 1. Go to the modules page and enable the contact.module.

How uncommon is for sites to have contact forms. This module should be enabled by default.

**Comment by klonos** [comment]
...baby steps :wink:

**Comment by jenlampton** [comment]
It looks like most the suggestions in this issue are listed in https://github.com/backdrop/backdrop-issues/issues/3023, so I'm tempted to close this one as duplicate in spite of it being created previously. I've added the following ides to that issue to make it complete.

- [ ] [UX] Allow contact form permissions to be set on the contact settings page
- [ ] [UX] Allow URL (path) to be set on the contact settings page
- [ ] Add a contact form context for Layouts
